### ISSUE TITLE
Eval Depth Performence

### ISSUE THREAD

**Issue by qihao-huang** [sconlyshootery/FeatDepth#17]
For KITTI Eigen Depth test by `fm_depth.pth`:

> -> Evaluating
using mean scaling
Scaling ratios | med: 37.704 | std: 0.061

> abs_rel| sq_rel| rmse| rmse_log| a1| a2| a3| 
&0.097 &0.676 &4.299 &0.172 &0.905 &0.967 &0.984 \\

-> Done!

The report I get is different from your paper. **Could you tell me which one it belongs to or anything wrong with my testing.**

@sconlyshootery 

<img width="768" alt="Screenshot 2020-11-21 at 00 14 52" src="https://user-images.githubusercontent.com/20598246/99822945-c37c8480-2b8e-11eb-8df8-e1b854fb4ec3.png">

**Comment by qihao-huang** [comment]
cfg_kitti_fm.py
`DEPTH_LAYERS = 50 #resnet50
POSE_LAYERS = 18 #resnet18
# FRAME_IDS = [0, -1, 1, 's'] 
FRAME_IDS = [0, -1, 1] 
#0 refers to current frame, -1 and 1 refer to temperally adjacent frames, 's' refers to stereo adjacent frame.
IMGS_PER_GPU = 1 #the number of images fed to each GPU
HEIGHT = 320 #input image height
WIDTH = 1024 #input image width

data = dict(
    name = 'kitti', #dataset name
    split = 'exp', #training split name
    height = HEIGHT,
    width = WIDTH,
    frame_ids = FRAME_IDS,
    in_path = '/home/qhhuang/monodepth-project/dataset/kitti_data', #path to raw data
    gt_depth_path = '/home/qhhuang/monodepth-project/monodepth2/splits/eigen/gt_depths.npz', #path to gt data
    png = True, #image format
    stereo_scale = True if 's' in FRAME_IDS else False,
)

model = dict(
    name = 'mono_fm',# select a model by name
    depth_num_layers = DEPTH_LAYERS,
    pose_num_layers = POSE_LAYERS,
    frame_ids = FRAME_IDS,
    imgs_per_gpu = IMGS_PER_GPU,
    height = HEIGHT,
    width = WIDTH,
    scales = [0, 1, 2, 3],# output different scales of depth maps
    min_depth = 0.1, # minimum of predicted depth value
    max_depth = 100.0, # maximum of predicted depth value
    depth_pretrained_path = '/home/qhhuang/monodepth-project/tmp/FeatDepth/resnet{}.pth'.format(DEPTH_LAYERS),
    pose_pretrained_path =  '/home/qhhuang/monodepth-project/tmp/FeatDepth/resnet{}.pth'.format(POSE_LAYERS),
    extractor_pretrained_path = '/home/qhhuang/monodepth-project/tmp/FeatDepth/autoencoder.pth',

    automask = False if 's' in FRAME_IDS else True,
    disp_norm = False if 's' in FRAME_IDS else True,
    perception_weight = 1e-3,
    smoothness_weight = 1e-3,
)

# resume_from = '/node01_data5/monodepth2-test/model/ms/ms.pth'#directly start training from provide weights
resume_from = None
finetune = None
total_epochs = 40
imgs_per_gpu = IMGS_PER_GPU
learning_rate = 1e-4
workers_per_gpu = 4
validate = True

optimizer = dict(type='Adam', lr=learning_rate, weight_decay=0)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[20,30],
    gamma=0.5,
)

checkpoint_config = dict(interval=1)
log_config = dict(interval=50,
                  hooks=[dict(type='TextLoggerHook'),])
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
workflow = [('train', 1)]`

**Comment by qihao-huang** [comment]
eval_depth.py
```
from __future__ import absolute_import, division, print_function
import cv2
import sys
import numpy as np
from mmcv import Config

import torch
from torch.utils.data import DataLoader

sys.path.append('.')
from mono.model.registry import MONO
from mono.model.mono_baseline.layers import disp_to_depth
from mono.datasets.utils import readlines, compute_errors
from mono.datasets.kitti_dataset import KITTIRAWDataset

cv2.setNumThreads(0)  # This speeds up evaluation 5x on our unix systems (OpenCV 3.3.1)
STEREO_SCALE_FACTOR = 36
MIN_DEPTH=1e-3
MAX_DEPTH=80

def evaluate(MODEL_PATH, CFG_PATH, GT_PATH):
    filenames = readlines("/home/qhhuang/monodepth-project/FeatDepth/mono/datasets/splits/exp/val_files.txt")
    cfg = Config.fromfile(CFG_PATH)

    dataset = KITTIRAWDataset(cfg.data['in_path'],
                              filenames,
                              cfg.data['height'],
                              cfg.data['width'],
                              [0],
                              is_train=False,
                              img_ext=".png",
                              gt_depth_path=GT_PATH)

    dataloader = DataLoader(dataset,
                            1,
                            shuffle=False,
                            num_workers=4,
                            pin_memory=True,
                            drop_last=False)

    cfg.model['imgs_per_gpu'] = 1
    model = MONO.module_dict[cfg.model['name']](cfg.model)
    checkpoint = torch.load(MODEL_PATH)
    model.load_state_dict(checkpoint['state_dict'], strict=True)
    model.cuda()
    model.eval()

    pred_disps = []
    with torch.no_grad():
        for batch_idx, inputs in enumerate(dataloader):
            for key, ipt in inputs.items():
                inputs[key] = ipt.cuda()
            outputs = model(inputs)

            disp = outputs[("disp", 0, 0)]

            pred_disp, _ = disp_to_depth(disp, 0.1, 100)
            pred_disp = pred_disp.cpu()[:, 0].numpy()
            pred_disps.append(pred_disp)
    pred_disps = np.concatenate(pred_disps)

    gt_depths = np.load(GT_PATH, allow_pickle=True, fix_imports=True, encoding='latin1')["data"]

    print("-> Evaluating")
    if cfg.data['stereo_scale']:
        print('using baseline')
    else:
        print('using mean scaling')

    errors = []
    ratios = []
    for i in range(pred_disps.shape[0]):
        gt_depth = gt_depths[i]
        gt_height, gt_width = gt_depth.shape[:2]

        pred_disp = pred_disps[i]
        pred_disp = cv2.resize(pred_disp, (gt_width, gt_height))

        pred_depth = 1 / pred_disp

        mask = np.logical_and(gt_depth > MIN_DEPTH, gt_depth < MAX_DEPTH)
        crop = np.array([0.40810811 * gt_height, 0.99189189 * gt_height,
                         0.03594771 * gt_width,  0.96405229 * gt_width]).astype(np.int32)
        crop_mask = np.zeros(mask.shape)
        crop_mask[crop[0]:crop[1], crop[2]:crop[3]] = 1
        mask = np.logical_and(mask, crop_mask)

        pred_depth = pred_depth[mask]
        gt_depth = gt_depth[mask]

        ratio = np.median(gt_depth) / np.median(pred_depth)
        ratios.append(ratio)

        if cfg.data['stereo_scale']:
            ratio = STEREO_SCALE_FACTOR

        pred_depth *= ratio
        pred_depth[pred_depth < MIN_DEPTH] = MIN_DEPTH
        pred_depth[pred_depth > MAX_DEPTH] = MAX_DEPTH
        errors.append(compute_errors(gt_depth, pred_depth))

    ratios = np.array(ratios)
    med = np.median(ratios)
    mean_errors = np.array(errors).mean(0)
    print("Scaling ratios | med: {:0.3f} | std: {:0.3f}".format(med, np.std(ratios / med)))
    print("\n" + ("{:>}| " * 7).format("abs_rel", "sq_rel", "rmse", "rmse_log", "a1", "a2", "a3"))
    print(("&{:.3f} " * 7).format(*mean_errors.tolist()) + "\\\\")
    print("\n-> Done!")


if __name__ == "__main__":
    CFG_PATH = '/home/qhhuang/monodepth-project/FeatDepth/config/cfg_kitti_fm.py'#path to cfg file
    GT_PATH = '/home/qhhuang/monodepth-project/monodepth2/splits/eigen/gt_depths.npz'#path to kitti gt depth
    MODEL_PATH = '/home/qhhuang/monodepth-project/tmp/FeatDepth/fm_depth.pth'#path to model weights
    evaluate(MODEL_PATH, CFG_PATH, GT_PATH)
```

**Comment by sconlyshootery** [comment]
The 'fm_depth.pth' corresponds with the numbers with the second to last row in table 2, it seems the performance you evaluted is much higher than  we report. You seem to incorrectly set stere_scale = False in the cfg file. We set stere_scale = False when we trained on monocular video in which case the model suffers from scale ambiguity , however the weights 'fm_depth.pth' is trained on binocular video with known baseline, so mean scale alignment shouldn't be used.

**Comment by qihao-huang** [comment]
Thanks for your reply!

It sounds like the provided model are trained with stereo assistance. May I ask do you have any schedule to release the monocular pretrained weights?

Thanks!

**Comment by sconlyshootery** [comment]
The provided model is alreadily trained in monocular manner, that is the input to the DepthNet is a single image. Trained on binocular videos indicates that we used predicted depth map to backproject to not only adjacent frames in tempral but also to its stereo pair to calculate photometric loss.

**Comment by qihao-huang** [comment]
Thanks for your Explanation.

I perform depth evaluation by setting  `FRAME_IDS = [0, -1, 1, 's']` instead of `FRAME_IDS = [0, -1, 1]` with your advice.
The tested model is downloaded by your google drive link.
Then, I get:

> Scaling ratios | med: 37.704 | std: 0.061
abs_rel| sq_rel| rmse| rmse_log| a1| a2| a3| 
&0.098 &0.697 &4.434 &0.184 &0.890 &0.963 &0.982 \\

The number is close to MS (trained on calibrated binocular videos).
<img width="702" alt="Screenshot 2020-11-24 at 22 44 59" src="https://user-images.githubusercontent.com/20598246/100109318-b02b2b00-2ea6-11eb-98fd-941f64a22f4f.png">

Given this observation, I assume the provided model is trained by binocular videos.
May I ask are there any mistakes or missing details? 

Thanks!

**Comment by sconlyshootery** [comment]
Congratulation! The new setting is correct, numbers are supposed to be exactly the same with the screenshot. But it also seems reasonable taking into account of different environment or some other details.
